File: android/app/src/main/kotlin/io/airtun/app/core/PinCode.kt
package io.airtun.app.core

import java.security.SecureRandom

/**
 * 4-digit numeric PIN for AirTun connection authentication and discovery pairing.
 * Generates a clean 4-digit numeric string (1000..9999) with no ambiguous chars.
 */
object PinCode {

    const val LENGTH = 4
    const val MIN = 1000
    const val MAX = 9999

    private val random = SecureRandom()

    fun draw(): String {
        return (MIN + random.nextInt(MAX - MIN + 1)).toString()
    }

    fun isValid(input: String?): Boolean {
        val normalized = normalize(input) ?: return false
        return normalized.length == LENGTH
    }

    fun normalize(input: String?): String? {
        if (input == null) return null
        val trimmed = input.filterNot { it.isWhitespace() }
        if (trimmed.length != LENGTH) return null
        if (!trimmed.all { it in '0'..'9' }) return null
        return trimmed
    }
}

File: windows/AirTun.Core/PinCode.cs
namespace AirTun.Core;

public static class PinCode
{
    public const int Length = 4;
    public const int Min = 1000;
    public const int Max = 9999;

    public static bool IsValid(string? input)
    {
        var normalized = Normalize(input);
        return normalized is not null && normalized.Length == Length;
    }

    public static string? Normalize(string? input)
    {
        if (string.IsNullOrWhiteSpace(input)) return null;
        var digits = new string(input.Where(ch => !char.IsWhiteSpace(ch)).ToArray());
        if (digits.Length != Length) return null;
        if (!digits.All(char.IsAsciiDigit)) return null;
        return digits;
    }
}
